### ISSUE TITLE
Form export fails under 11.5

### ISSUE THREAD

**Issue by AstCommodore** [frappant/frp_form_answers#53]
Hello guys!

Under 

- T3 11.5.13
- frp_form_answers 4.0.0 
- PHP 8.1 

the export fails or better explained:

The form is submitted correctly, no error occurs (also all logs are fine), but the data isn't saved. It doesn't matter, if the finisher is the first in the form configuration or not. see also #47 
I don't have a hint actually, what could go wrong here ...

Another problem occurs in the Backend in the extension module:

> PHP Warning
> Core: Error handler (BE): PHP Warning: Undefined array key "tot" in /var/www/html/public/typo3conf/ext/frp_form_answers/Classes/Utility/FormAnswersUtility.php line 49

The FormAnswersUtility loops throw the entries and counts by

`$pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'] += 1;`

Maybe it's better to check if there is an array key:

```
foreach ($startPointPids as $pageId) {
            foreach ($this->formEntryRepository->findAllInPidAndRootline($pageId) as $formEntry) {
                if (isset($pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'])) {
                    $pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'] += 1;
                } else {
                    $pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'] = 1;
                }

                if (!$formEntry->isExported()) {
                    if (isset($pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'])) {
                        $pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'] += 1;
                    } else {
                        $pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'] = 1;
                    }
                }
            }
        }
```

**Comment by AstCommodore** [comment]
I have to add 

protected PersistenceManager $persistenceManager;

```
    public function injectPersistenceManager(PersistenceManager $persistenceManager): void
    {
        $this->persistenceManager = $persistenceManager;
    }
```

and 

```
$this->formEntryRepository->add($this->formEntry);
$this->persistenceManager->persistAll();

```

to the SaveFormToDatabaseFinisher – now the finisher saves again!

**Comment by AstCommodore** [comment]
I wrote some composer patches targeting different section all over the extension to get it working under TYPO3 11.5 – maybe you want to use them to modify the code:

```
--- a/Resources/Private/Language/de.locallang_be.xlf	2022-04-07 15:48:58.000000000 +0200
+++ b/Resources/Private/Language/de.locallang_be.xlf	2022-07-27 16:28:19.000000000 +0200
@@ -37,7 +37,7 @@
             </trans-unit>
 			<trans-unit id="tx_frpformanswers_backend_templates_formentry_show.delete">
 				<source>Delete entries</source>
-				<target>Eingaben löschen</target>
+				<target>Einträge löschen</target>
 			</trans-unit>
 			<trans-unit id="flashmessage.deleteFormName.header">
 				<source>Entries deleted</source>
@@ -141,7 +141,7 @@
             </trans-unit>
             <trans-unit id="tx_frpformanswers_backend_templates_formentry_list.nothingsaved">
                 <source>No saved formEntries</source>
-                <target>Keine gespeicherten formEinträge</target>
+                <target>Keine gespeicherten Einträge</target>
             </trans-unit>
 		</body>
 	</file>

```

```
--- a/Classes/Utility/FormAnswersUtility.php	2022-04-07 15:48:58.000000000 +0200
+++ b/Classes/Utility/FormAnswersUtility.php	2022-07-27 15:57:28.000000000 +0200
@@ -3,6 +3,7 @@

 use Frappant\FrpFormAnswers\Domain\Repository\FormEntryRepository;
 use TYPO3\CMS\Core\Domain\Repository\PageRepository;
+use TYPO3\CMS\Core\Utility\DebugUtility;
 use TYPO3\CMS\Extbase\Utility\DebuggerUtility;

 class FormAnswersUtility
@@ -44,10 +45,18 @@
         // Get all Pids with a formEntry list
         foreach ($startPointPids as $pageId) {
             foreach ($this->formEntryRepository->findAllInPidAndRootline($pageId) as $formEntry) {
-                $pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'] += 1;
+                if (isset($pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'])) {
+                    $pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'] += 1;
+                } else {
+                    $pageIds[$formEntry->getPid()][$formEntry->getForm()]['tot'] = 1;
+                }

                 if (!$formEntry->isExported()) {
-                    $pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'] += 1;
+                    if (isset($pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'])) {
+                        $pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'] += 1;
+                    } else {
+                        $pageIds[$formEntry->getPid()][$formEntry->getForm()]['new'] = 1;
+                    }
                 }
             }
         }

```

```
--- a/Classes/Controller/FormEntryController.php	2022-04-07 15:48:58.000000000 +0200
+++ b/Classes/Controller/FormEntryController.php	2022-07-28 11:44:22.000000000 +0200
@@ -19,6 +19,7 @@
 use TYPO3\CMS\Extbase\Mvc\Controller\ActionController;
 use TYPO3\CMS\Extbase\Utility\LocalizationUtility;
 use TYPO3\CMS\Extbase\Utility\DebuggerUtility;
+use TYPO3\CMS\Core\Domain\Repository\PageRepository;

 /***
  *
@@ -62,6 +63,16 @@
     protected DataExporter $dataExporter;

     /**
+     * @var PageRepository $pageRepository
+     */
+    protected PageRepository $pageRepository;
+
+    /**
+     * @var integer
+     */
+    protected $pid;
+
+    /**
      * http headers to send with filedownload request @see exportAction
      *
      * @var array
@@ -74,13 +85,17 @@
         IconFactory $iconFactory,
         FormAnswersUtility $formAnswersUtility,
         FormEntryRepository $formEntryRepository,
-        DataExporter $dataExporter
+        DataExporter $dataExporter,
+        PageRepository $pageRepository
     ) {
         $this->moduleTemplateFactory = $moduleTemplateFactory;
         $this->iconFactory = $iconFactory;
         $this->formAnswersUtility = $formAnswersUtility;
         $this->formEntryRepository = $formEntryRepository;
         $this->dataExporter = $dataExporter;
+        $this->pageRepository = $pageRepository;
+
+        $this->pid = (int)GeneralUtility::_GP('id');
     }

     /**
@@ -136,7 +151,7 @@

         $count = $queryBuilder->count('*')
             ->from('tx_frpformanswers_domain_model_formentry')
-            ->where($queryBuilder->expr()->eq('pid', $queryBuilder->createNamedParameter(\Frappant\FrpFormAnswers\Utility\BackendUtility::getCurrentPid(), \PDO::PARAM_INT)))
+            ->where($queryBuilder->expr()->eq('pid', $queryBuilder->createNamedParameter($this->pid, \PDO::PARAM_INT)))
             ->andWhere($queryBuilder->expr()->eq('deleted', $queryBuilder->createNamedParameter(1, \PDO::PARAM_INT)))
             ->execute()->fetchFirstColumn();
         //DebuggerUtility::var_dump($count);
@@ -161,12 +176,12 @@
         $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('tx_frpformanswers_domain_model_formentry');

         $queryBuilder->delete('tx_frpformanswers_domain_model_formentry')
-            ->where($queryBuilder->expr()->eq('pid', $queryBuilder->createNamedParameter(\Frappant\FrpFormAnswers\Utility\BackendUtility::getCurrentPid(), \PDO::PARAM_INT)))
+            ->where($queryBuilder->expr()->eq('pid', $queryBuilder->createNamedParameter($this->pid, \PDO::PARAM_INT)))
             ->andWhere($queryBuilder->expr()->eq('deleted', $queryBuilder->createNamedParameter(1, \PDO::PARAM_INT)))
             ->execute();

         $this->addFlashMessage(
-            LocalizationUtility::translate('LLL:EXT:frp_form_answers/Resources/Private/Language/de.locallang_be.xlf:flashmessage.removeEntries.body', null, [\Frappant\FrpFormAnswers\Utility\BackendUtility::getCurrentPid()]),
+            LocalizationUtility::translate('LLL:EXT:frp_form_answers/Resources/Private/Language/de.locallang_be.xlf:flashmessage.removeEntries.body', null, [$this->pid]),
             LocalizationUtility::translate('LLL:EXT:frp_form_answers/Resources/Private/Language/de.locallang_be.xlf:flashmessage.removeEntries.title'),
             \TYPO3\CMS\Core\Messaging\FlashMessage::OK,
             true);
@@ -314,11 +329,11 @@
             $queryBuilder->update(
                 'tx_frpformanswers_domain_model_formentry',
                 [ 'deleted' => 1 ], // set
-                [ 'form' => $formName, 'pid' => \Frappant\FrpFormAnswers\Utility\BackendUtility::getCurrentPid()]
+                [ 'form' => $formName, 'pid' => $this->pid]
             );

             $this->addFlashMessage(
-                LocalizationUtility::translate('LLL:EXT:frp_form_answers/Resources/Private/Language/de.locallang_be.xlf:flashmessage.deleteFormName.body', 'frp_form_answers', [$formName, \Frappant\FrpFormAnswers\Utility\BackendUtility::getCurrentPid()]),
+                LocalizationUtility::translate('LLL:EXT:frp_form_answers/Resources/Private/Language/de.locallang_be.xlf:flashmessage.deleteFormName.body', 'frp_form_answers', [$formName, $this->pid]),
                 LocalizationUtility::translate('LLL:EXT:frp_form_answers/Resources/Private/Language/de.locallang_be.xlf:flashmessage.deleteFormName.title'),
                 \TYPO3\CMS\Core\Messaging\FlashMessage::OK,
                 true);

```

```
--- a/Resources/Private/Backend/Templates/FormEntry/List.html	2022-04-07 15:48:58.000000000 +0200
+++ b/Resources/Private/Backend/Templates/FormEntry/List.html	2022-07-27 16:07:21.000000000 +0200
@@ -48,8 +48,8 @@
                 </f:link.action>
             </div>
         </f:form>
-        <f:be.tableList tableName="tx_frpformanswers_domain_model_formentry" filter="{formName}" enableControlPanels="true"
-                        fieldList="{0:'uid', 1: 'crdate'}" storagePid="{pid}" clickTitleMode="edit"/>
+        <f:be.tableList tableName="tx_frpformanswers_domain_model_formentry" filter="{formName}" enableClickMenu="true" enableControlPanels="true"
+                        fieldList="{0:'uid', 1: 'form', 2: 'crdate', 3: 'exported'}" storagePid="{pid}" clickTitleMode="edit"/>
     </f:for>

     <f:if condition="{subPagesWithFormEntries}">


```

```
--- a/Classes/Domain/Finishers/SaveFormToDatabaseFinisher.php	2022-04-07 15:48:58.000000000 +0200
+++ b/Classes/Domain/Finishers/SaveFormToDatabaseFinisher.php	2022-07-27 17:07:18.000000000 +0200
@@ -4,6 +4,8 @@
 use Frappant\FrpFormAnswers\Event\ManipulateFormValuesEvent;
 use Frappant\FrpFormAnswers\Domain\Model\FormEntry;
 use TYPO3\CMS\Core\EventDispatcher\EventDispatcher;
+use TYPO3\CMS\Core\Utility\DebugUtility;
+use TYPO3\CMS\Extbase\Persistence\Generic\PersistenceManager;
 use TYPO3\CMS\Extbase\Utility\DebuggerUtility;
 use TYPO3\CMS\Form\Domain\Finishers;
 use TYPO3\CMS\Core\Database\ConnectionPool;
@@ -59,6 +61,13 @@
         $this->formEntry = $formEntry;
     }

+    protected PersistenceManager $persistenceManager;
+
+    public function injectPersistenceManager(PersistenceManager $persistenceManager): void
+    {
+        $this->persistenceManager = $persistenceManager;
+    }
+
     /**
      * Executes this finisher
      * @see AbstractFinisher::execute()
@@ -103,6 +112,7 @@
         $this->formEntry->setSubmitUid($lastFormUid);

         $this->formEntryRepository->add($this->formEntry);
+        $this->persistenceManager->persistAll();
     }

     /**

```

**Comment by frpuser** [comment]
Hey @AstCommodore 
Thank you for your diligent help in improving frp-form-answers.
I have adopted all your improvements and released them in version 4.0.1.

Best regards
!Frappant
